extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTaskClient.java

```diff
@@ -297,11 +265,31 @@ private StatusResponseHolder submitRequest(Request request, boolean retry)
         } else if (responseCode == 400) { // don't bother retrying if it's a bad request
           throw new IAE("Received 400 Bad Request with body: %s", response.getContent());
         } else {
-          throw new IOException(String.format("Received status [%d] with: %s", responseCode, response.getContent()));
+          throw new IOException(String.format("Received status [%d]", responseCode));
         }
       }
       catch (IOException | ChannelException e) {
-        final Duration delay = retryPolicy.getAndIncrementRetryDelay();
+
+        // Since workers are free to move tasks around to different ports, there is a chance that a task may have been
+        // moved but our view of its location has not been updated yet from ZK. To detect this case, we send a header
+        // identifying our expected recipient in the request; if this doesn't correspond to the worker we messaged, the
+        // worker will return an HTTP 404 with its ID in the response header. If we get a mismatching task ID, then
+        // we will wait for a short period then retry the request indefinitely, expecting the task's location to
+        // eventually be updated.
+
+        final Duration delay;
+        if (response != null && response.getStatus().equals(HttpResponseStatus.NOT_FOUND)) {
```

Comment:
trying indefinitely scares me a little bit, what happens if the task's location is never updated?